### PR TITLE
Shared model/new query executor

### PR DESCRIPTION
**Pull request by x3medima17**
### Description of the Change
This PR migrates query execution to the new model completely.

### Benefits
Old query execution pipeline removed

### Possible Drawbacks 
None

### CODE REVIEW

File: shared_model/builders/protobuf/builder_templates/query_response_template.hpp [hunk 1]
```diff
@@ -89,16 +89,22 @@ namespace shared_model {
       auto accountAssetResponse(
           const interface::types::AssetIdType &asset_id,
           const interface::types::AccountIdType &account_id,
-          const std::string &amount) const {
+          const interface::Amount &amount) const {
```
Comment (muratovv) [inline shared_model/builders/protobuf/builder_templates/query_response_template.hpp]: What is the reason for the changes?

File: test/module/irohad/validation/query_execution.cpp [hunk 2]
```diff
@@ -40,6 +41,7 @@
 #include "model/queries/responses/signatories_response.hpp"
```
Comment (muratovv) [inline test/module/irohad/validation/query_execution.cpp]: Have to rework all includes here.

File: test/module/irohad/validation/query_execution.cpp [hunk 3]
```diff
@@ -54,6 +56,9 @@ using namespace framework::test_subscriber;
 
 using wTransaction = std::shared_ptr<shared_model::interface::Transaction>;
 
+template <class T>
+using w = shared_model::detail::PolymorphicWrapper<T>;
```
Comment (muratovv) [inline test/module/irohad/validation/query_execution.cpp]: Poly wrapper is not allowed to use here.

File: test/module/irohad/validation/query_execution.cpp [hunk 4]
```diff
@@ -474,14 +532,19 @@ TEST_F(GetAccountAssetsTest, DifferentDomainAccountInValidCase) {
...
 
-  auto response = validateAndExecute();
-  auto cast_resp = std::static_pointer_cast<ErrorResponse>(response);
-  ASSERT_EQ(cast_resp->reason, ErrorResponse::STATEFUL_INVALID);
+  auto response = validateAndExecute(query);
+
+  auto cast_resp = boost::get<w<shared_model::interface::ErrorQueryResponse>>(
+      response->get());
+
+  ASSERT_NO_THROW(
```
Comment (muratovv) [inline test/module/irohad/validation/query_execution.cpp]: `ASSERT_NO_THROW` with a cast of the variant inside is very common. Maybe better to move it into own function?

File: test/module/irohad/validation/query_execution.cpp [hunk 5]
```diff
@@ -647,37 +747,39 @@ TEST_F(GetSignatoriesTest, DifferentDomainAccountInValidCase) {
...
   role_permissions = {can_get_all_signatories};
 
   EXPECT_CALL(*wsv_query, getAccountRoles(admin_id))
       .WillOnce(Return(admin_roles));
   EXPECT_CALL(*wsv_query, getRolePermissions(admin_role))
       .WillOnce(Return(role_permissions));
 
-  EXPECT_CALL(*wsv_query, getSignatories(get_signatories->account_id))
-      .WillOnce(Return(boost::none));
+  EXPECT_CALL(*wsv_query, getSignatories("none")).WillOnce(Return(boost::none));
```
Comment (muratovv) [inline test/module/irohad/validation/query_execution.cpp]: Why getSignatorites has been reworked with "none" instead of `accountId()`.

File: irohad/model/impl/query_execution.cpp [hunk 6]
```diff
@@ -76,374 +91,402 @@ bool hasQueryPermission(const std::string &creator,
...
+  return clone(response);
+}
+
+std::shared_ptr<shared_model::interface::QueryResponse>
+QueryProcessingFactory::execute(const shared_model::interface::Query &query) {
+  const auto &query_hash = query.hash();
+  return visit_in_place(
+      query.get(),
+      [&](const w<shared_model::interface::GetAccount> &q) {
+        if (not validate(query, *q)) {
```
Comment (nickaleks) [inline irohad/model/impl/query_execution.cpp]: Code duplication below, consider making a separate function

File: test/module/irohad/validation/query_execution.cpp [hunk 7]
```diff
@@ -696,13 +803,14 @@ TEST_F(GetAccountTransactionsTest, MyAccountValidCase) {
...
-  auto TxWrapper = make_test_subscriber<CallExact>(txs_observable, N);
-  TxWrapper.subscribe(
-      [this](auto val) { EXPECT_EQ(admin_id, val->creatorAccountId()); });
-  ASSERT_TRUE(TxWrapper.validate());
+  auto response = validateAndExecute(query);
+  auto cast_resp = boost::get<w<shared_model::interface::TransactionsResponse>>(
+      response->get());
+
+  for (const auto &tx : cast_resp->transactions()) {
+    EXPECT_EQ(admin_id, tx->creatorAccountId());
```
Comment (muratovv) [inline test/module/irohad/validation/query_execution.cpp]: Looks like suspicious changes: the previous version change the internal invariant and number of calls. This version checks only invariant.

File: irohad/model/impl/query_execution.cpp [hunk 8]
```diff
@@ -16,6 +16,11 @@
  */
 
 #include "model/query_execution.hpp"
+#include <boost/algorithm/string.hpp>
+#include <rxcpp/rx-observable.hpp>
```
Comment (nickaleks) [inline irohad/model/impl/query_execution.cpp]: Order includes according to the style guide pls

**Review by nickaleks — CHANGES_REQUESTED** [review]

File: test/module/irohad/validation/query_execution.cpp [hunk 9]
```diff
@@ -745,12 +863,14 @@ TEST_F(GetAccountTransactionsTest, DomainAccountValidCase) {
...
 
-  auto TxWrapper = make_test_subscriber<CallExact>(txs_observable, N);
-  TxWrapper.subscribe(
-      [this](auto val) { EXPECT_EQ(account_id, val->creatorAccountId()); });
-  ASSERT_TRUE(TxWrapper.validate());
+  auto response = validateAndExecute(query);
+  auto cast_resp = boost::get<w<shared_model::interface::TransactionsResponse>>(
+      response->get());
+
+  for (const auto &tx : cast_resp->transactions()) {
```
Comment (muratovv) [inline test/module/irohad/validation/query_execution.cpp]: same as before

File: test/module/irohad/validation/query_execution.cpp [hunk 10]
```diff
@@ -759,26 +879,32 @@ TEST_F(GetAccountTransactionsTest, DomainAccountValidCase) {
...
```
Comment (muratovv) [inline test/module/irohad/validation/query_execution.cpp]: Sab

**Review by muratovv — CHANGES_REQUESTED** [review]

File: irohad/model/impl/query_execution.cpp [hunk 11]
```diff
@@ -29,13 +34,15 @@
...
 
-#include <boost/algorithm/string.hpp>
-#include <rxcpp/rx-observable.hpp>
-#include <utility>
-
 using namespace iroha::model;
 using namespace iroha::ametsuchi;
 
+template <class T>
+using w = shared_model::detail::PolymorphicWrapper<T>;
```
[the remark this answers is not in the document]
Comment (kamilsa) [reply irohad/model/impl/query_execution.cpp]: We decided to remove it in https://soramitsu.atlassian.net/secure/RapidBoard.jspa?rapidView=9&view=planning&selectedIssue=IR-1011
@x3medima17 please, leave TODO with IR-1011 to fix that later

File: shared_model/builders/protobuf/builder_templates/query_response_template.hpp (hunk 1 above)
Comment (x3medima17) [reply shared_model/builders/protobuf/builder_templates/query_response_template.hpp]: Less error prone, also it is better to pass concrete object, instead of one of its representation

File: irohad/model/impl/query_execution.cpp [hunk 12]
```diff
@@ -76,374 +91,402 @@ bool hasQueryPermission(const std::string &creator,
...
-  auto ast = shared_ast | [&](auto &asset) {
-    return boost::make_optional(*std::unique_ptr<iroha::model::Asset>(asset->makeOldModel()));
-  };
+std::shared_ptr<shared_model::interface::QueryResponse>
+QueryProcessingFactory::executeGetAssetInfo(
+    const shared_model::interface::GetAssetInfo &query,
+    const shared_model::interface::types::HashType &hash) {
+  auto ast = _wsvQuery->getAsset(query.assetId());
+
   if (not ast) {
```
[the remark this answers is not in the document]
Comment (x3medima17) [reply irohad/model/impl/query_execution.cpp]: `ast` is optional, but method returns `shared_ptr`

File: irohad/model/impl/query_execution.cpp [hunk 13]
```diff
@@ -76,374 +91,402 @@ bool hasQueryPermission(const std::string &creator,
...
-    return std::make_shared<ErrorResponse>(response);
+    return build_error<shared_model::interface::NoAccountErrorResponse>(hash);
   }
 
-  AccountResponse response;
-  response.account = acc.value();
-  response.roles = roles.value();
-  response.query_hash = iroha::hash(query);
-  return std::make_shared<AccountResponse>(response);
+  auto account = std::static_pointer_cast<shared_model::proto::Account>(*acc);
```
[the remark this answers is not in the document]
Comment (x3medima17) [reply irohad/model/impl/query_execution.cpp]: I must pass `proto::Account` since it uses `.getTransport()` method in the builder

File: irohad/model/impl/query_execution.cpp [hunk 14]
```diff
@@ -76,374 +91,402 @@ bool hasQueryPermission(const std::string &creator,
...
-  response.query_hash = iroha::hash(query);
-  response.transactions = acc_asset_tx.map([](const auto &tx) {
-    return *std::unique_ptr<iroha::model::Transaction>(tx->makeOldModel());
-  });
-  return std::make_shared<TransactionsResponse>(response);
+    const shared_model::interface::GetAccountAssetTransactions &query,
+    const shared_model::interface::types::HashType &hash) {
+  auto acc_asset_tx = _blockQuery->getAccountAssetTransactions(
+      query.accountId(), query.assetId());
+
```
[the remark this answers is not in the document]
Comment (x3medima17) [reply irohad/model/impl/query_execution.cpp]: Changed `reduce` to `subscribe`

File: test/module/irohad/validation/query_execution.cpp (hunk 5 above)
Comment (x3medima17) [reply test/module/irohad/validation/query_execution.cpp]: `account_id` was proto field, it was set in constructor and then changed in every test, and it was hard to trace all changes, so I removed it, also because we use builders now

File: test/module/irohad/validation/query_execution.cpp (hunk 7 above)
Comment (x3medima17) [reply test/module/irohad/validation/query_execution.cpp]: Now we don't have observable, in response we have only raw vectors

File: test/module/irohad/validation/query_execution.cpp (hunk 9 above)
Comment (x3medima17) [reply test/module/irohad/validation/query_execution.cpp]: same as before 
:D

File: test/module/irohad/validation/query_execution.cpp (hunk 10 above)
Comment (x3medima17) [reply test/module/irohad/validation/query_execution.cpp]: sab

File: test/module/irohad/validation/query_execution.cpp (hunk 4 above)
Comment (x3medima17) [reply test/module/irohad/validation/query_execution.cpp]: It is 2 lines of code, the function would have more, and we have to pass the response as parameter and the error as template, so it is an overkill

**Review by nickaleks — APPROVED** [review]

File: irohad/model/query_execution.hpp [hunk 15]
```diff
@@ -31,22 +31,37 @@
 #include "ametsuchi/block_query.hpp"
 #include "ametsuchi/wsv_query.hpp"
 
+#include "builders/protobuf/builder_templates/query_response_template.hpp"
```
Comment (muratovv) [inline irohad/model/query_execution.hpp]: Remove model includes

File: test/module/irohad/validation/query_execution.cpp (hunk 7 above)
Comment (muratovv) [reply test/module/irohad/validation/query_execution.cpp]: Well, this means that you miss the number of calls (same as a number of elements) invariant.

File: test/module/irohad/validation/query_execution.cpp (hunk 9 above)
Comment (muratovv) [reply test/module/irohad/validation/query_execution.cpp]: =(

File: test/module/irohad/validation/query_execution.cpp [hunk 16]
```diff
@@ -884,12 +1014,14 @@ TEST_F(GetAccountAssetsTransactionsTest, AllAccountValidCase) {
...
-  auto response = validateAndExecute();
 
-  auto TxWrapper = make_test_subscriber<CallExact>(txs_observable, N);
-  TxWrapper.subscribe(
-      [this](auto val) { EXPECT_EQ(account_id, val->creatorAccountId()); });
-  ASSERT_TRUE(TxWrapper.validate());
+  auto response = validateAndExecute(query);
+  auto cast_resp = boost::get<w<shared_model::interface::TransactionsResponse>>(
+      response->get());
+
```
Comment (muratovv) [inline test/module/irohad/validation/query_execution.cpp]: I think here also need to make a number elements check.

**Review by muratovv — CHANGES_REQUESTED** [review]

**Review by muratovv — APPROVED** [review]

File: irohad/model/impl/query_execution.cpp [hunk 17]
```diff
@@ -76,374 +90,344 @@ bool hasQueryPermission(const std::string &creator,
...
-  }
-  ErrorResponse response;
-  response.query_hash = iroha::hash(*query);
-  response.reason = model::ErrorResponse::NOT_SUPPORTED;
-  return std::make_shared<ErrorResponse>(response);
+std::shared_ptr<shared_model::interface::QueryResponse>
+QueryProcessingFactory::execute(const shared_model::interface::Query &query) {
+  const auto &query_hash = query.hash();
+  QueryResponseBuilderDone builder;
+  return visit_in_place(
```
Comment (lebdron) [inline irohad/model/impl/query_execution.cpp]: Please add TODO with task id to rewrite this as in https://github.com/hyperledger/iroha/pull/1007/files#diff-cdb3f6acf132ab4b63459a816a9a50b5R80 , so that code is not duplicated.
